File: typescript/leetcodehard/regularExpressionMatching.ts
/*
https://leetcode.com/problems/regular-expression-matching/

status - completed
*/

import { logp } from "../utils/logger";

function isMatch(s: string, p: string): boolean {
  const m = s.length;
  const n = p.length;
  const dp: boolean[][] = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(false));
  dp[0][0] = true;

  for (let j = 1; j <= n; j++) {
    if (p[j - 1] === "*") {
      dp[0][j] = dp[0][j - 2];
    }
  }

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (p[j - 1] === "*") {
        dp[i][j] = dp[i][j - 2];
        const prev = p[j - 2];
        if (prev === "." || prev === s[i - 1]) {
          dp[i][j] = dp[i][j] || dp[i - 1][j];
        }
      } else if (p[j - 1] === "." || p[j - 1] === s[i - 1]) {
        dp[i][j] = dp[i - 1][j - 1];
      }
    }
  }

  return dp[m][n];
}

logp(`input s="aa" p="a" expected false actual ${isMatch("aa", "a")}`);
logp(`input s="aa" p="a*" expected true actual ${isMatch("aa", "a*")}`);
logp(`input s="mississippi" p="mis*is*p*." expected false actual ${isMatch("mississippi", "mis*is*p*.")}`);

File: python/leetcodehard/regular_expression_matching.py
"""
https://leetcode.com/problems/regular-expression-matching/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def isMatch(self, s, p):
        memo = {}

        def dp(i, j):
            if (i, j) in memo:
                return memo[(i, j)]

            if j == len(p):
                result = i == len(s)
            else:
                first_match = i < len(s) and p[j] in (s[i], ".")

                if j + 1 < len(p) and p[j + 1] == "*":
                    result = dp(i, j + 2) or (first_match and dp(i + 1, j))
                else:
                    result = first_match and dp(i + 1, j + 1)

            memo[(i, j)] = result
            return result

        return dp(0, 0)


if __name__ == "__main__":
    test_input_1 = ("aa", "a*")
    output = Solution().isMatch(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ("mississippi", "mis*is*p*.")
    output = Solution().isMatch(*test_input_2)
    Solution().log.info("Output is %s", output)
